client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -91,26 +88,107 @@ func NewMessageService(ip string, port int, me types.Address, pk []byte) *P2PMes
 	if err != nil {
 		panic(err)
 	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ms.cancel = cancel
+
+	mdns := mdns.NewMdnsService(host, "", ms)
+	if err := mdns.Start(); err != nil {
+		panic(err)
+	}
+	ms.mdns = mdns
 	ms.p2pHost = host
 
-	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, func(stream network.Stream) {
-		reader := bufio.NewReader(stream)
-		// Create a buffer stream for non blocking read and write.
-		raw, err := reader.ReadString(DELIMITER)
+	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, ms.msgStreamHandler)
+	ms.p2pHost.SetStreamHandler(PEER_EXCHANGE_PROTOCOL_ID, ms.gossipStreamHandler)
+
+	go ms.gossip(ctx)
+
+	return ms
+}
+
+// HandlePeerFound is called by the mDNS service when a peer is found.
+func (ms *P2PMessageService) HandlePeerFound(pi peer.AddrInfo) {
+	fmt.Printf("Found peer %v\n", pi)
+	ms.p2pHost.Peerstore().AddAddr(pi.ID, pi.Addrs[0], peerstore.PermanentAddrTTL)
+}
+
+func (ms *P2PMessageService) msgStreamHandler(stream network.Stream) {
+	reader := bufio.NewReader(stream)
+	// Create a buffer stream for non blocking read and write.
+	raw, err := reader.ReadString(DELIMITER)
 
-		// An EOF means the stream has been closed by the other side.
-		if errors.Is(err, io.EOF) {
-			stream.Close()
+	// An EOF means the stream has been closed by the other side.
+	if errors.Is(err, io.EOF) {
+		stream.Close()
+		return
+	}
+	ms.checkError(err)
+	m, err := protocols.DeserializeMessage(raw)
+	ms.checkError(err)
+	ms.toEngine <- m
+	stream.Close()
+}
+
+func (ms *P2PMessageService) gossipStreamHandler(stream network.Stream) {
```

Comment:
If we were ok enforcing that the `pk` for the  message service is your statechannel key, we could probably derive the state channel address from the `peer.ID`. This means we wouldn't manually have to gossip our state channel address to our peers.